libs/vscode-ui/components/src/lib/autocomplete/autocomplete.component.ts

```diff
@@ -1,16 +1,225 @@
+  isFocused$: Observable<boolean>;
+
+  private readonly _destroying$ = new Subject<void>();
+
+  constructor(private readonly _elementRef: ElementRef) {}
+
+  writeValue(value: string) {
+    if (this.control) {
+      this.control.setValue(value);
+    } else {
+      this.control = new FormControl(value);
+      this.visibleOptions = this._options$.pipe(
+        switchMap(options =>
+          this.control.valueChanges.pipe(
+            startWith(this.control.value),
+            map(formValue =>
+              options.filter(option =>
+                option.toLowerCase().includes(formValue.toLowerCase())
+              )
+            )
+          )
+        )
+      );
+    }
+  }
+
+  ngOnInit() {
+    this.isFocused$ = merge(
+      fromEvent(this._elementRef.nativeElement, 'focusin').pipe(
+        map(() => 'focus in')
+      ),
+      fromEvent(this._elementRef.nativeElement, 'focusout').pipe(
+        map(() => 'focus out')
+      )
+    ).pipe(
+      map(last => (last === 'focus in' ? true : false)),
+      startWith(false),
+      debounceTime(300),
+      shareReplay(1)
+    );
+
+    this.selectedOption$ = this.isFocused$.pipe(
+      switchMap(isFocused =>
+        isFocused
+          ? this.visibleOptions.pipe(
+              switchMap(visibleOptions =>
+                visibleOptions.length
+                  ? fromEvent<KeyboardEvent>(document, 'keyup').pipe(
+                      map((event: KeyboardEvent) => event.key),
+                      filter(key => this.isNavigationKey(key)),
+                      scan(
+                        ([index, _]: [number, string | null], key) =>
+                          [
+                            this.updatedOptionIndex(
+                              <AutocompleteNavKeys>key,
+                              index || 0,
+                              visibleOptions.length
+                            ),
+                            key
+                          ] as [number, string | null],
+                        [0, null] as [number, string | null]
+                      ),
+                      startWith([0, null] as [number, string | null]),
+
+                      tap(([index, key]) => {
+                        if (key === AutocompleteNavKeys.Enter) {
+                          this.optionSelected(visibleOptions[index]);
+                        }
+                      }),
+
+                      map(([selectedIndex]) => selectedIndex)
+                    )
+                  : of(null)
+              )
+            )
+          : of(null)
+      ),
+      shareReplay(1)
+    );
+
+    this.selectedOption$.subscribe();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.options) {
+      this._options$.next(this.options);
+    }
+  }
+
+  /**
+   * Check if the input key is one relevant to autocomplete panel navigation
+   */
+  isNavigationKey = (key: KeyboardEvent['key']): boolean => {
+    return (
+      key === AutocompleteNavKeys.ArrowUp ||
```

Comment:
It would be slightly more efficient and idiomatic to define a module level constant of the et of AutocompleteNavKeys this function cares about

Reply:
Done